Allow Docker daemon to bind to different IP than --cluster-advertise.

yongtang opened the pull request:
**\- What I did**

This fix intends to address the issue raised in #22008 and #22087 where Docker's overlay networking can only bind to the IP specified in the option of `--cluster-advertise`. In certain situations, end user may want the overlay networking to bind to an internal IP while at the same time advertise a publicly accessible IP.

**\- How I did it**

This fix add an additional option of --cluster-listen to allow passing a differnt IP than --cluster-advertise.

**\- How to verify it**

Additional tests to cover the changes have been added.

In AWS,
1. Create an EC2 instances with IP of x.x.x.x
2. Assign an EIP y.y.y.y to the created instance.
3. Invoke docker daemon with
    `
    dockerd --debug --cluster-store=consul://<consul>:8500 --cluster-advertise=y.y.y.y:2376 --cluster-listen=0.0.0.0:2376
`
4. Run a docker daemon from another host and point the `--cluster-store` to the same key-value store.
5. Follow the same steps as specified in docs to create multi-host networking:
https://docs.docker.com/engine/userguide/networking/get-started-overlay/

**\- Description for the changelog**

Add the flag --cluster-listen for overlay network so that it will listen to a different IP than the advertised one (specified in --cluster-advertise).

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #22008. This fix fixes #22087.

**NOTE: Separate pull requests in libnetwork and engine-api:**
https://github.com/docker/libnetwork/pull/1175 
https://github.com/docker/engine-api/pull/365

Signed-off-by: Yong Tang yong.tang.github@outlook.com

thaJeztah commented:
ping @vieux, @mavenugo @cpuguy83 could you have a look at this?

mavenugo commented:
design LGTM. Moving to code-review.

Please open a PR against docker/libnetwork for all libnetwork specific changes.

yongtang commented:
Thanks @mavenugo for the feedback. I created a pull request in libnetwork:
https://github.com/docker/libnetwork/pull/1175
Please let me know if there are any issues.

cpuguy83 commented:
Needs a rebase and vendoring done.

yongtang committed:
Allow Docker daemon to bind to different IP than --cluster-advertise.

This fix intends to address the issue raised in 22008 and 22087 where
Docker's overlay networking can only bind to the IP specified in the
option of `--cluster-advertise`. In certain situations, end user may
want the overlay networking to bind to an internal IP while at the same
time advertise a publicly accessible IP.

This fix add an additional option of --cluster-listen to allow passing
a differnt IP than --cluster-advertise.

Additional tests to cover the changes have been added.

This fix fixes #22008. This fix fixes #22087.

NOTE: Separate pull requests in engine-api and libnetwork  will be created
accordingly.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @cpuguy83. The pull request has been updated. I also opened a pull request in engine-api:
https://github.com/docker/engine-api/pull/365

Please let me know if there are any issues.

cpuguy83 commented:
This needs a rebase. Also engine-api is now moved back to this repo so any changes will need to be made there.

yongtang committed:
Placeholder for vendoring libnetwork.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @cpuguy83. The pull request has been updated and it does not require engine-api any more. It still requires libnetwork though.

yongtang commented:
Thanks @thaJeztah. The PR has been rebased and updated.

stevvooe commented:
Could a cluster listen on multiple addresses?

yongtang commented:
@stevvooe In the non-swarm mode overlay networks, Serf was used. It only allows to specify a bind address and a advertise address. So I think it will be very difficult to find a way to listen to multiple address (unless it is 0.0.0.0).

yongtang commented:
it looks like there is not enough movement for this PR. Also with swarm mode overlay network in place, the old non-swarm overlay network might be phased out from docker in the future. Close this PR for now.
